recipes/gn/all/conanfile.py

```diff
@@ -0,0 +1,122 @@
+
+    def configure(self):
+        if self.settings.compiler.cppstd:
+            tools.check_min_cppstd(self, 17)
+        else:
+            if self._minimum_compiler_version_supporting_cxx17:
+                if tools.Version(self.settings.compiler.version) < self._minimum_compiler_version_supporting_cxx17:
+                    raise ConanInvalidConfiguration("gn requires a compiler supporting c++17")
+            else:
+                self.output.warn("gn recipe does not recognize the compiler. gn requires a compiler supporting c++17. Assuming it does.")
+
+    def package_id(self):
+        del self.info.settings.compiler
+
+    def source(self):
+        tools.get(**self.conan_data["sources"][self.version], destination=self._source_subfolder)
+
+    def build_requirements(self):
+        self.build_requires("ninja/1.10.2")
+
+    @contextmanager
+    def _build_context(self):
+        if self.settings.compiler == "Visual Studio":
+            with tools.vcvars(self.settings):
+                yield
+        else:
+            compiler_defaults = {}
+            if self.settings.compiler == "gcc":
+                compiler_defaults = {
+                    "CC": "gcc",
+                    "CXX": "g++",
+                    "AR": "ar",
+                    "LD": "g++",
+                }
+            elif self.settings.compiler == "clang":
+                compiler_defaults = {
+                    "CC": "clang",
+                    "CXX": "clang++",
+                    "AR": "ar",
+                    "LD": "clang++",
+                }
+            env = {}
+            for k in ("CC", "CXX", "AR", "LD"):
+                v = tools.get_env(k, compiler_defaults.get(k, None))
+                if v:
+                    env[k] = v
+            with tools.environment_append(env):
+                yield
+
+    @staticmethod
+    def _to_gn_platform(os_, compiler):
+        if tools.is_apple_os(os_):
+            return "darwin"
+        if compiler == "Visual Studio":
+            return "msvc"
+        # Assume gn knows about the os
+        return str(os_).lower()
+
+    def build(self):
+        with tools.chdir(self._source_subfolder):
+            with self._build_context():
+                tools.save(os.path.join("src", "gn", "last_commit_position.h"),
+                           textwrap.dedent("""\
+                                #pragma once
+                                #define LAST_COMMIT_POSITION "1"
+                                #define LAST_COMMIT_POSITION_NUM 1
+                                """))
+                conf_args = [
+                    "--no-last-commit-position",
+                    "--host={}".format(self._to_gn_platform(self.settings.os, self.settings.compiler)),
+                ]
+                if self.settings.build_type == "Debug":
+                    conf_args.append("-d")
+                self.run("python build/gen.py {}".format(" ".join(conf_args)), run_environment=True)
```

Comment:
Would it require cpython in build requirement when packaged? Maybe add a TODO?

Reply:
Yes, I'll add a todo.